[RFC] pm: device power management improvements

gmarull opened the pull request:
This RFC PR is an initial attempt to simplify the current device PM callback and fix some other issues. Even though the change set is large (all drivers have been ported), the behavior _should_ be the same as before. Only a few existing issues have been fixed, but there are many that still need to be addressed. As a result of these changes, adding support for device PM should be less cumbersome.

The list of changes can be summarized to:
1. Local power states are removed, device PM holds the device state
2. The "ctrl_command" is gone (`uint32_t ctrl_command`), devices are now called with the action they should perform: `static int my_device_pm_control(const struct device *dev, enum pm_device_action action)`.
3. Device PM control function is only called if needed. This means that devices are only called if the requested action will change the current state (e.g. a suspended device will not be called with `PM_DEVICE_ACTION_SUSPEND`).
4. `PM_DEVICE_STATE_FORCE_SUSPEND` has been removed since it did not represent any actual state and was not used.
5. `PM_DEVICE_STATE_LOW_POWER` has been replaced for `PM_DEVICE_STATE_SUSPEND`, since they were equivalents.
6. Device PM control function return codes are now documented.
7. Multiple implementation cleanups

**NOTE: This PR should only be taken as a reference, smaller changes will be submitted as new PRs.**

gmarull committed:
pm: device: remove usage of local states

The device PM subsystem already holds the device state, so there is no
need to keep duplicates inside the device. The pm_device_state_get has
been refactored to just return the device state. Note that this is still
not safe, but the same applied to the previous implementation. This
problem will be addressed later.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove usage of ctrl_command

The callback is now invoked to set the device PM state in all cases, so
the usage of ctrl_command is redundant.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove ctrl_command callback argument

The ctrl_command is not used anymore, so remove it from the callback
signature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: pm: update documentation

Update documentation so that latest changes are reflected.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove pointer usage for state

Since the state is no longer modified by the device PM callback, just
use the state value.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: improve docs and arg names for state set

Improve the docstrings of the pm_device_state_set function and change to
shorter argument variable names.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented on doc/reference/power_management/index.rst:
```diff
@@ -330,20 +325,17 @@ Device Set Power State
 .. code-block:: c
 
-   int pm_device_state_set(const struct device *dev, uint32_t device_power_state, pm_device_cb cb, void *arg);
+   int pm_device_state_set(const struct device *dev, enum pm_device_state *state);
```
The signature is wrong. State is not a pointer

gmarull replied:
fixed

ceolin commented on doc/reference/power_management/index.rst:
```diff
@@ -330,20 +325,17 @@ Device Set Power State
-
-Calls the :c:func:`pm_control()` handler function implemented by the
-device driver with PM_DEVICE_STATE_GET command.
+   int pm_device_state_get(const struct device *dev, enum pm_device_state *state);
```
ditto

gmarull replied:
here it is a pointer

ceolin reviewed (commented):
Removing local power states and the ctrl_command is really good. I'm
don't like removing the pm_device_cb without an alternative. Removing it without a replacement means that there we'll not have a way to do async operation.
Well ... unless the driver returns a transition state like RESUMING or SUSPENDING and then the runtime API checks it, but then the runtime will need to somehow
polling this value because the driver is not able to notify when the transition finishes.

gmarull commented:
> don't like removing the pm_device_cb without an alternative. Removing it without a replacement means that there we'll not have a way to do async operation.
> Well ... unless the driver returns a transition state like RESUMING or SUSPENDING and then the runtime API checks it, but then the runtime will need to somehow
> polling this value because the driver is not able to notify when the transition finishes.

If the `pm_control` function could be asynchronous, it should probably return `-EAGAIN` and the driver would later notify the device PM subsystem, e.g. using a new device PM API call. I'm not sure what alternative should be provided, since the current callback doesn't do anything related to asynchronous behavior. In any case, IMO this needs to be removed from every driver:

```
       if (cb) {
               cb(dev, ret, state, arg);
       }
```

It's not clearly documented, ugly and even in some cases it was being misused for non-PM stuff (see `samples/sensor/fdc2x1x/src/main.c`).

ceolin commented:
> > don't like removing the pm_device_cb without an alternative. Removing it without a replacement means that there we'll not have a way to do async operation.
> > Well ... unless the driver returns a transition state like RESUMING or SUSPENDING and then the runtime API checks it, but then the runtime will need to somehow
> > polling this value because the driver is not able to notify when the transition finishes.
> 
> If the `pm_control` function could be asynchronous, it should probably return `-EAGAIN` and the driver would later notify the device PM subsystem, e.g. using a new device PM API call. I'm not sure what alternative should be provided, since the current callback doesn't do anything related to asynchronous behavior. In any case, IMO this needs to be removed from every driver:
> 
> ```
>        if (cb) {
>                cb(dev, ret, state, arg);
>        }
> ```
> 
> It's not clearly documented, doesn't solve async problem, is ugly and even in some cases it was being misused for non-PM stuff (see `samples/sensor/fdc2x1x/src/main.c`).

I'm not defending the current API. It's ugly and need to be changed. What I'm saying is to have something before doing it. And the async path can be done with this callback. It is a matter of the device do call it later. The runtime device API waits this callback to broadcast the conditional variable. What I mean is, a driver can return from this function while the device is initialized and later call this callback. This should work with the runtime API.

Again, I don't love i. This API is confuse, the names are bad, blah blah blah. But I think we should not remove it without a proper replacement.

gmarull commented:
> I'm not defending the current API. It's ugly and need to be changed. What I'm saying is to have something before doing it. And the async path can be done with this callback. It is a matter of the device do call it later. The runtime device API waits this callback to broadcast the conditional variable. What I mean is, a driver can return from this function while the device is initialized and later call this callback. This should work with the runtime API.

It's true that a driver can store the callback and arg for later call (again, a bad design pattern). In this case, the device would be left in "suspending state" when using device runtime PM until it "notifies back". However, device PM doesn't support such behavior as far as I know, meaning async behavior can't be safely implemented as of today.

> 
> Again, I don't love i. This API is confuse, the names are bad, blah blah blah. But I think we should not remove it without a proper replacement.

These changes should not be incompatible with a future proper solution. I think it's better to start with a simple solution that works other than patching behaviors that are actually not used or not documented and that will likely require core changes to make them useful.

ceolin commented:
> > I'm not defending the current API. It's ugly and need to be changed. What I'm saying is to have something before doing it. And the async path can be done with this callback. It is a matter of the device do call it later. The runtime device API waits this callback to broadcast the conditional variable. What I mean is, a driver can return from this function while the device is initialized and later call this callback. This should work with the runtime API.
> 
> It's true that a driver can store the callback and arg for later call (again, a bad design pattern). In this case, the device would be left in "suspending state" when using device runtime PM until it "notifies back". However, device PM doesn't support such behavior as far as I know, meaning async behavior can't be safely implemented as of today.

Agreed, no doubts about bad design. By device PM you are referring when the system suspend or resume a device ? or is it something else ?

> 
> > Again, I don't love i. This API is confuse, the names are bad, blah blah blah. But I think we should not remove it without a proper replacement.
> 
> These changes should not be incompatible with a future proper solution. I think it's better to start with a simple solution that works other than patching behaviors that are actually not used or not documented and that will likely require core changes to make them useful.

They are not incompatible, my problem is that we don't have any plan when this functionality will come in and how it will be. Anyway, I won't block this change because of this, but I really don't like changes that breaks the functionality without some replacement.

ceolin commented:
@gmarull would you mind to create a new pr with the commit changing the devices states to enum type ? We can get that one merged while we continue to discuss this one.

gmarull commented:
> @gmarull would you mind to create a new pr with the commit changing the devices states to enum type ? We can get that one merged while we continue to discuss this one.

While the commits are _atomic_, I'd prefer to discuss all the changes first and merge them all together. I think it's better to merge the _final_ callback instead of going through multiple refactors.

I'd feel more comfortable to merge this sort of changes early to avoid the situation of 2.6, when non-trivial changes were merged in the last minute.

The point that I'd like to discuss in particular, since it implies a callback change, is:

```
The callback may operate using "actions" and not "states", e.g. PM_DEVICE_ACTION_SUSPEND,
which would lead to the device in PM_DEVICE_STATE_SUSPENDED. States such as
PM_DEVICE_STATE_SUSPENDING or PM_DEVICE_STATE_RESUMING do not represent actual
callback actions.
```

ceolin reviewed (commented):
That is my point since the beginning. This *probably* is not the final version until we have decided how to properly handle
asynchronous operations.

gmarull committed:
pm: device: do not call device if already on the given state

If the device is already at the given state, do not call the device PM
control function. This makes sure that devices are only called to change
from one state to another.

Even though asynchronous device PM is completely broken, transitional
states are considered too.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: simplify state change check logic

The device PM control function will only be called if the requested
state is different from the current one. A significant amount of drivers
were checking for state changes, now unnecessary. This patch removes all
this redundant logic.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: remove unused force suspend state

In a first place, `PM_DEVICE_STATE_FORCE_SUSPEND` does not represent a
state but an action (force the suspension of a device). It turns out
that it was not being used for anything related to state suspension.
Finally, the state lacked proper documentation. Since the behavior
"claimed" by this state is not implemented at all, remove it entirely
from the tree.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: replace low power state with suspend state

The `PM_DEVICE_STATE_LOW_POWER` was being used as an equivalent to the
`PM_DEVICE_STATE_SUSPEND`. However, it had some problems: no support
for transitional states (ie no possibility of async), not supported by
the runtime API, etc. Since it does not bring any additional value, just
remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: rename PM_DEVICE_STATE_SUSPEND to PM_DEVICE_STATE_SUSPENDED

The verb tense for the suspended state was not consistent with other
states.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: cleanup pm control callback implementations

- Return -ENOTSUP if the requested state is not supported
- Remove redundant "noop style" functions.
- Use switch everywhere to handle requested state

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: fix incorrect usages of PM_DEVICE_STATE_OFF

According to the documentation the OFF state has to be used when the
devices is fully turned off, ie, power removed. Most drivers were using
a sort of fall-through for all non-active states, leading to behaviors
not following the specifications.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: add type for device PM control callback

Add a type and documentation for the device PM control callback. This
way possible return codes and its meaning are documented.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: use actions for device PM control

Instead of passing target states, use actions for device PM control.
Actions represent better the meaning of the callback argument.
Furthermore, they are more future proof as they can be suitable for
other PM actions that have no direct mapping to a state. If we compare
with Linux, we could have a multi-stage suspend/resume. Such scenario
would not have a good mapping when using target states.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: led: led_pwm: remove non-compliant PM

The driver was manually setting the state of each PWM LED when the
device PM control function was invoked. This is not what the device PM
control function is supposed to do. If the PWM device has support for
device PM, it will be invoked without further work from the driver. If
this behavior is expected to be happening at runtime, Zephyr does not
currently support managing power dependencies automatically.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
@ceolin @nashif The rationale behind _pm actions_:

If we take Linux as a reference, all PM operations are held by `struct dev_pm_ops`:
```c
struct dev_pm_ops {
	...
	int (*prepare)(struct device *dev);
	void (*complete)(struct device *dev);
	int (*suspend)(struct device *dev);
	int (*resume)(struct device *dev);
        ...
};
```
ref https://elixir.bootlin.com/linux/latest/source/include/linux/pm.h#L277

As you can see, each callback represents an _action_: _prepare_, _suspend_... States are not reflected here. The proposal for Zephyr is to have a single callback (as we do now) with the action as an argument, that is:
```c
int my_pm_control(const struct device *dev, enum pm_device_action action);
```

When drawing a device state diagram, actions become the natural way to represent transitions:

![image](https://user-images.githubusercontent.com/25011557/125058995-22a1f180-e0ab-11eb-9009-54da4adcdd0c.png)

If considering async case:
![image](https://user-images.githubusercontent.com/25011557/125058038-25501700-e0aa-11eb-9379-93ae762855ec.png)

ceolin commented:
> @ceolin @nashif The rationale behind _pm actions_:
> 
> If we take Linux as a reference, all PM operations are held by `struct dev_pm_ops`:
> 
> ```c
> struct dev_pm_ops {
> 	...
> 	int (*prepare)(struct device *dev);
> 	void (*complete)(struct device *dev);
> 	int (*suspend)(struct device *dev);
> 	int (*resume)(struct device *dev);
>         ...
> };
> ```
> 
> ref https://elixir.bootlin.com/linux/latest/source/include/linux/pm.h#L277
> 
> As you can see, each callback represents an _action_: _prepare_, _suspend_... States are not reflected here. The proposal for Zephyr is to have a single callback (as we do now) with the action as an argument, that is:
> 
> ```c
> int my_pm_control(const struct device *dev, enum pm_device_action action);
> ```
> 
> When drawing a device state diagram, actions become the natural way to represent transitions:
> 
> ![image](https://user-images.githubusercontent.com/25011557/125063229-83332d80-e0af-11eb-970e-c34890c77742.png)
> 
> If considering async case:
> ![image](https://user-images.githubusercontent.com/25011557/125063410-baa1da00-e0af-11eb-8d54-ca0959d09dcb.png)

I liked the idea of using actions instead of states.

nashif reviewed (changes requested):
need to discuss PM_DEVICE_STATE_LOW_POWER removal.

ceolin commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1831,14 +1827,14 @@ static void uarte_nrfx_set_power_state(const struct device *dev,
 				       enum pm_device_state new_state)
 {
 	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+#if defined(CONFIG_UART_ASYNC_API) || defined(UARTE_INTERRUPT_DRIVEN)
```
Is this part of this change ?

gmarull replied:
yes, previous to the removal of local states pm_control used `data` unconditionally (to get state, which was always available).

gmarull commented:
Closing, #36759 contains relevant changes listed here.
